kong/plugins/rate-limiting/handler.lua

```diff
@@ -64,6 +68,32 @@ local function get_usage(conf, api_id, identifier, current_timestamp, limits)
   return usage, stop
 end
 
+local function get_period_end_in_secs(current_timestamp, limits)
+  local timetable
+  local periods = timestamp.get_timestamps(current_timestamp)
+
+  for period, period_date in pairs(periods) do
+    if limits[period] then
+      timetable = new_from_timestamp(floor(period_date / 1000))
+
+      -- Unfortunately 'second' & 'minute' are 'sec' and 'min' in timetable
+      if period == "second" then
+        timetable.sec = timetable.sec + 1
+
+      elseif period == "minute" then
+        timetable.min = timetable.min + 1
+
+      else
+        timetable[period] = timetable[period] + 1
+      end
+    end
+  end
+
+  -- Mutates the current object's time and date components so that are integers within 'normal' ranges
+  timetable:normalise()
+  return timetable:timestamp()
+end
```

Comment:
Is it me, or is this logic flawed? I see a few issues with it as best as I can tell.

Let's consider the following configuration:
```json
{
    "config.second": 10,
    "config.minute": 20,
    "config.hour": 100
}
```

1. Now, what this loop does is declaring a new `timetable` variable *inside of the loop*. This means than for the above example, the loop will run 3 times, and each time the `timetable` variable will have be mutated, but all mutations are dismissed as soon as the next iteration kicks-in, because we override the variable (See L. 77).

2. Additionally, this uses `pairs()`. Which means the order on which we will iterate over these values is un-deterministic. This means that if we the current time `t` is `01:01:01` (`HH:mm:ss`), our reset time `t_reset` could be: `01:01:02`, assuming the `second` period was the last one yielded by `pairs()`. This would be problematic because if a Consumer made 30 requests in a minute, `t_reset` would definitely not be `01:01:02`.

3. Now, suppose we reuse the same `timetable` variable in this loop but keep the same logic. Suppose `t` is `01:01:01` again. This logic would return a `t_reset` of `02:02:02`, which is not the correct time at which the rate-limiting will reset. The correct time is: `02:00:00`.

I think this logic should be reworked: instead, we should find the biggest period that the user has configured in the rate-limiting plugin (from the above example, that would be `config.hour`), and increment this unit of time by `1` (to switch to the next fixed time window of that period). This would return a `t_reset` of `02:00:00`.

---

**Side note:** for a Consumer, I am sure that knowing the time at which the rates will reset is valuable. However, quid of the time at which the Consumer will be allowed to make requests again? This does seem more valuable to me. Example:

Considering the same rate-limiting configuration:
```json
{
    "config.second": 10,
    "config.minute": 20,
    "config.hour": 100
}
```

Finally, consider a Consumer who made the following requests at `01:00:xx`:
  - `01:00:00`: 8
  - `01:00:01`: 8
  - `01:00:02`: 4

Now, that Consumer is denied from making more requests to that API because they are exceeding the `config.minute` window of `20`. This patch would inform the Consumer that their rate-limiting will reset on the next hour, `02:00:00`. But the user will be able to make request earlier than that. The next request the Consumer will be able to make will be at `t_retry` of `01:01:00`. Quite a lot sooner than `t_reset`.

As a Consumer, isn't `t_retry` more valuable than knowing `t_reset`?

Reply:
I concur, the logic needs to be fixed -- I completely overlooked that, sorry. As for `retry`, I'd say it's useful to have both the complete reset timestamp and the fine-grained retry? @grahamar are you willing to work on this?

Reply:
@grahamar ping, any update?